src/contract.rs

```diff
@@ -1,25 +1,33 @@
 use cosmwasm_std::{
-    attr, entry_point, to_binary, Addr, Binary, BlockInfo, Coin, CosmosMsg, Deps, DepsMut, Env,
-    MessageInfo, Reply, Response, StdError, StdResult, SubMsg, Uint128, WasmMsg,
+    attr, entry_point, to_binary, Addr, Binary, BlockInfo, Coin, CosmosMsg, Decimal, Deps, DepsMut,
+    Env, MessageInfo, Reply, Response, StdError, StdResult, SubMsg, Uint128, Uint256, Uint512,
+    WasmMsg,
 };
 use cw0::parse_reply_instantiate_data;
 use cw2::set_contract_version;
 use cw20::Denom::Cw20;
 use cw20::{Cw20ExecuteMsg, Denom, Expiration, MinterResponse};
 use cw20_base::contract::query_balance;
+use std::convert::TryInto;
+use std::str::FromStr;
 
 use crate::error::ContractError;
 use crate::msg::{
     ExecuteMsg, InfoResponse, InstantiateMsg, QueryMsg, Token1ForToken2PriceResponse,
     Token2ForToken1PriceResponse, TokenSelect,
 };
-use crate::state::{Token, LP_TOKEN, TOKEN1, TOKEN2};
+use crate::state::{Fees, Token, FEES, LP_TOKEN, OWNER, TOKEN1, TOKEN2};
 
 // Version info for migration info
 pub const CONTRACT_NAME: &str = "crates.io:wasmswap";
 pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
 const INSTANTIATE_LP_TOKEN_REPLY_ID: u64 = 0;
+
+const FEE_SCALE_FACTOR: Uint128 = Uint128::new(10_000);
+const MAX_FEE_PERCENT: &str = "1";
+const FEE_DECIMAL_PRECISION: Uint128 = Uint128::new(100_000_000_000_000_000_000);
```

Comment:
Could you help me understand where this number comes from? This math precision stuff can get weird.

Reply:
So the Decimal type has [default decimals](https://docs.rs/cosmwasm-std/1.0.0-beta/cosmwasm_std/struct.Decimal.html#method.decimal_places) of 18, which is constant. Two additional zeros are added to account for the percents being expressed as out of 100, instead of 1.

Reply:
Ugh this one too. Didn't mean the submit the review in the old state. Correct comment now in the edited version.

Reply:
Ahh gotcha. Yeah that's clean. Will revise

Reply:
So the problem with this is that decimal_places requires an instance of decimal, it's not static. And even if you used Decimal::zero() or Decimal::one(), decimal_places is a non-const function so can't use it when defining a const variable. I'm thinking of just using 

```
Uint128::new(10u128.pow(20))
```
Should be fine since the decimal type scale is constant for now. Can add a comment explaining this